diff --git a/CMakeLists.txt b/CMakeLists.txt
index acf8e35..be5dbf3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -3,6 +3,9 @@ cmake_minimum_required(VERSION 3.14)
 project(zsign)
 set(CMAKE_CXX_STANDARD 11)
 
+#https://stackoverflow.com/questions/51121295/cmake-find-module-to-distinguish-shared-or-static-library
+set(CMAKE_FIND_LIBRARY_SUFFIXES ".a" ".so")
+
 # Dependencies
 # On macOS, search Homebrew for keg-only versions of OpenSSL because system provided /usr/lib/libssl.dylib cannot be linked
 if (CMAKE_HOST_SYSTEM_NAME MATCHES "Darwin")
@@ -27,6 +30,7 @@ message("OpenSSL libraries: ${OPENSSL_LIBRARIES}")
 include_directories(${OPENSSL_INCLUDE_DIR})
 list(APPEND LIB_LIST ${OPENSSL_LIBRARIES})
 
+
 find_package(ZLIB REQUIRED)
 message("zlib include dir: ${ZLIB_INCLUDE_DIR}")
 message("zlib libraries: ${ZLIB_LIBRARIES}")
@@ -39,5 +43,13 @@ list(APPEND LIB_LIST ${ZLIB_LIBRARIES})
 file(GLOB SRC common/*.cpp ./**.cpp)
 
 add_executable(zsign ${SRC})
-target_link_libraries(zsign ${LIB_LIST})
-
+target_link_options(zsign PRIVATE -static-libgcc -static-libstdc++)
+#target_link_libraries(zsign ${LIB_LIST})
+
+set(OPENSSL_USE_STATIC_LIBS TRUE)
+find_package(OpenSSL REQUIRED)
+target_link_libraries(zsign PRIVATE OpenSSL::Crypto)
+find_package(Threads REQUIRED)
+target_link_libraries(zsign PRIVATE Threads::Threads)
+find_package(ZLIB REQUIRED)
+target_link_libraries(zsign PRIVATE ZLIB::ZLIB)
